### DISCUSSION TITLE
Cache partitioning

### DISCUSSION THREAD

**Discussion by Everaldo-Gomes** [bao-project/bao-hypervisor#140] [General]
Hi!
Can I get some guidance about how to configure / use cache partitioning in the hypervisor?

Platform: Raspberry Pi 4b
Demo: Dual-guest Linux+FreeRTOS

**Comment by Diogo21Costa** [comment]
Hi @Everaldo-Gomes,

The cache coloring can be configured using the `colors` field within the `vm_config` [struct](https://github.com/bao-project/bao-hypervisor/blob/main/src/core/inc/config.h#L88).

This field corresponds to a bitmap for the assigned cache colors of the VM. This value is truncated depending on the number of available colors calculated at run-time, i.e., it is platform-dependent. By default, the coloring mechanism is not active. 

For instance, in your setup, you could assign 50% of cache colors to each VM by setting `.colors=0b00001111` to Linux and `.colors=0b11110000` to FreeRTOS.

Please feel free to let me know if you need any more help.

**Reply by Everaldo-Gomes** [reply]
To get this working, the

```
size_t COLOR_NUM = 1;
size_t COLOR_SIZE = 1;
```

in the [cache.c](https://github.com/bao-project/bao-hypervisor/blob/main/src/core/cache.c) should be changed?
In this case, would `COLOR_NUM = 2` be sufficient?

**Reply by Diogo21Costa** [reply]
You don't need to, all you need to do is to update your configuration file (e.g., if you're using the [config file from bao demos](https://github.com/bao-project/bao-demos/blob/master/demos/linux%2Bfreertos/configs/rpi4.c) you could add the following changes:

```diff
        { 
            .image = {
                .base_addr = 0x20000000,
                .load_addr = VM_IMAGE_OFFSET(linux_image),
                .size = VM_IMAGE_SIZE(linux_image)
            },

            .entry = 0x20000000,
+         .colors = 0b11110000,

            .platform = {
```

```diff
            .image = {
                .base_addr = 0x0,
                .load_addr = VM_IMAGE_OFFSET(freertos_image),
                .size = VM_IMAGE_SIZE(freertos_image)
            },

            .entry = 0x0,
+         .colors = 0b00001111,

            .platform = {
```

**Reply by Everaldo-Gomes** [reply]
Thank you so much!

**Reply by Everaldo-Gomes** [reply]
"_Assign 50% of cache colors to each VM by setting .colors=0b00001111 to Linux and .colors=0b11110000 to FreeRTOS._"
I believe it means that an amount of colors is already defined.

How can I define the number of colors? 
Where can I find how the .colors=0b00001111 is transformed/mapped in the partitions?

Are these values related to 32 or 64 bits, and how do I know this?

**Reply by Diogo21Costa** [reply]
The allocation of cache colors to each VM is predetermined based on the architecture of the cache system. The number of cache partitions is determined by the structure of the cache, typically calculated as the way size divided by the page size. To confirm the number of colors available for a specific platform, one can check the variable `COLOR_NUM`, typically calculated at runtime for ARM-based platforms (please check this [file](https://github.com/bao-project/bao-hypervisor/blob/main/src/core/cache.c)).


Regarding the values specified in the assignment (`0b00001111` for Linux and `0b11110000` for FreeRTOS), these are bitmasks indicating which cache colors each VM can utilize, where each bit corresponds to one color. These values are platform-specific and related to the size of the cache way (my example was for a zcu104). However, for the Raspberry Pi 4 platform, there are only four colors available. Therefore, for a 50/50 cache coloring scheme, the bitmask for Linux would be `0b1100`, and for FreeRTOS, it would be `0b0011`.

**Reply by Everaldo-Gomes** [reply]
Thanks for the detailed explanation, I appreciate it!
